### PR TITLE
[no sq] Server bug fixes and unit tests

### PR DESCRIPTION
**Pull request by sfan5**
## To do

This PR is a ~~Work in Progress~~ / Ready for Review.

## How to test

* [commit 1] see it working [here](https://github.com/sfan5/minetest/actions/runs/8115484923/job/22183446522)
* [commit 2] no known reproducer, apply eyes to code
* [commit 3] manual testing:
  * test code from #14420
  * can also test the code from #14311
  * **note**: the new unit tests do exactly this
* [commit 4] manual testing:
  * use the test code from #14420 and a revision *before* commit 4 to create a world with statically saved objects that shouldn't be static
  * join world once, objects should appear
  * join world for the second time, they should be gone
  * **note**: the new unit tests do exactly this
* do some general testing with animal mods or whatever to make sure I didn't mess up

### CODE REVIEW

**Comment by grorp** [comment]
It would be very nice if you could add unittests for c566ffebc51226f44697e76a5476ab0aa3a04710 and 8e7f3b97032d61d7d36809f094b9d0ab26c977d1. The behavior seems easy to mess up and (at least in parts) well testable.

**Comment by sfan5** [comment]
I'll check how easily I can test it from C++. Devtest is challenging due to missing direct control and insight into this part.

**Comment by rubenwardy** [comment]
I have tested that this fixes the menuclouds assertion error

**Comment by sfan5** [comment]
@grorp Tests added.

File: src/server.cpp [hunk 1]
```diff
@@ -369,12 +381,10 @@ Server::~Server()
...
 		m_env->saveMeta();
-	}
 
-	// Stop threads
-	if (m_thread) {
-		stop();
-		delete m_thread;
+		// Note that this also deletes and saves the map.
+		delete m_env;
+		m_env = nullptr;
```
Comment (nerzhul) [inline src/server.cpp:387]: setting m_env to nullptr in destructor means we have a scope issue somewhere in the destruction. Which code part uses m_env after that and will break ? (it means we should find a wait to split some objects at a moment)

File: src/serverenvironment.cpp [hunk 2]
```diff
@@ -1882,16 +1882,42 @@ void ServerEnvironment::getSelectedActiveObjects(
...
 
+	// Register reference in scripting api (must be done before post-init)
+	m_script->addObjectReference(object);
+	// Post-initialize object
+	// Note that this can change the value of isStaticAllowed() in case of LuaEntitySAO
+	object->addedToEnvironment(dtime_s);
+
+	// Activate object
+	if (object->m_static_exists)
+	{
```
Comment (nerzhul) [inline src/serverenvironment.cpp:1900]: wrong brace :)

File: src/serverenvironment.cpp [hunk 3]
```diff
@@ -2049,18 +2064,26 @@ void ServerEnvironment::activateObjects(MapBlock *block, u32 dtime_s)
...
-		verbosestream << "ServerEnvironment::activateObjects(): "
-			<< "activated static object pos=" << (s_obj.pos / BS)
-			<< " type=" << (int)s_obj.type << std::endl;
+
+		obj->m_static_exists = true;
+		obj->m_static_block = block->getPos();
+
 		// This will also add the object to the active static list
-		addActiveObjectRaw(std::move(obj), false, dtime_s);
+		bool ok = addActiveObjectRaw(std::move(obj), &s_obj, dtime_s) != 0;
```
Comment (nerzhul) [inline src/serverenvironment.cpp:2079]: you can assign & check in the condition directly

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: src/server.cpp (hunk 1 above)
Comment (sfan5) [reply src/server.cpp:387]: this is just good practice to not leave dangling pointers around

File: src/serverenvironment.cpp (hunk 2 above)
Comment (sfan5) [reply src/serverenvironment.cpp:1900]: intentional so I can properly place the comment before `else`, suggestions welcome

File: src/serverenvironment.cpp (hunk 3 above)
Comment (sfan5) [reply src/serverenvironment.cpp:2079]: that's possible but I don't see the advantage

**Review by Desour — APPROVED** [review]
I'm not too familiar with static object stuff. But code looks sane.

**Pull request merged**